std/arith.asm

```diff
@@ -17,18 +23,176 @@ machine Arith(CLK32_31, operation_id){
+    // Equal to 0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2f
+    let secp_modulus = (1 << 256) - (1 << 32) - (1 << 9) - (1 << 8) - (1 << 7) - (1 << 6) - (1 << 4) - 1;
+
+    let inverse: int -> int = |x| ff::inverse(x, secp_modulus);
+    let add = |x, y| ff::add(x, y, secp_modulus);
+    let sub = |x, y| ff::sub(x, y, secp_modulus);
+    let mul = |x, y| ff::mul(x, y, secp_modulus);
+    let div = |x, y| ff::div(x, y, secp_modulus);
+
+    // TODO: Somehow this doesn't work in the queries?
+    // "Expected array for std::array::len, but got main_arith.x1: col"
+    // pol commit x1[16], y1[16], x2[16], y2[16], x3[16], y3[16];
+    pol commit x1__0, x1__1, x1__2, x1__3, x1__4, x1__5, x1__6, x1__7, x1__8, x1__9, x1__10, x1__11, x1__12, x1__13, x1__14, x1__15;
+    let x1 = [x1__0, x1__1, x1__2, x1__3, x1__4, x1__5, x1__6, x1__7, x1__8, x1__9, x1__10, x1__11, x1__12, x1__13, x1__14, x1__15];
+    pol commit y1__0, y1__1, y1__2, y1__3, y1__4, y1__5, y1__6, y1__7, y1__8, y1__9, y1__10, y1__11, y1__12, y1__13, y1__14, y1__15;
+    let y1 = [y1__0, y1__1, y1__2, y1__3, y1__4, y1__5, y1__6, y1__7, y1__8, y1__9, y1__10, y1__11, y1__12, y1__13, y1__14, y1__15];
+    pol commit x2__0, x2__1, x2__2, x2__3, x2__4, x2__5, x2__6, x2__7, x2__8, x2__9, x2__10, x2__11, x2__12, x2__13, x2__14, x2__15;
+    let x2 = [x2__0, x2__1, x2__2, x2__3, x2__4, x2__5, x2__6, x2__7, x2__8, x2__9, x2__10, x2__11, x2__12, x2__13, x2__14, x2__15];
+    pol commit y2__0, y2__1, y2__2, y2__3, y2__4, y2__5, y2__6, y2__7, y2__8, y2__9, y2__10, y2__11, y2__12, y2__13, y2__14, y2__15;
+    let y2 = [y2__0, y2__1, y2__2, y2__3, y2__4, y2__5, y2__6, y2__7, y2__8, y2__9, y2__10, y2__11, y2__12, y2__13, y2__14, y2__15];
+    pol commit x3__0, x3__1, x3__2, x3__3, x3__4, x3__5, x3__6, x3__7, x3__8, x3__9, x3__10, x3__11, x3__12, x3__13, x3__14, x3__15;
+    let x3 = [x3__0, x3__1, x3__2, x3__3, x3__4, x3__5, x3__6, x3__7, x3__8, x3__9, x3__10, x3__11, x3__12, x3__13, x3__14, x3__15];
+    pol commit y3__0, y3__1, y3__2, y3__3, y3__4, y3__5, y3__6, y3__7, y3__8, y3__9, y3__10, y3__11, y3__12, y3__13, y3__14, y3__15;
+    let y3 = [y3__0, y3__1, y3__2, y3__3, y3__4, y3__5, y3__6, y3__7, y3__8, y3__9, y3__10, y3__11, y3__12, y3__13, y3__14, y3__15];
+
+    let s_for_eq1 = |x1, y1, x2, y2, limb_index| (div(sub(y2, y1), sub(x2, x1)) >> (limb_index * 16)) & 0xffff;
+    let s_for_eq2 = |x1, y1, limb_index| (div(mul(3, mul(x1, x1)), mul(2, y1)) >> (limb_index * 16)) & 0xffff;
+
+    let compute_x3_int = |x1, x2, s| (s * s - x1 - x2 + 2 * secp_modulus) % secp_modulus;
+    let compute_y3_int = |x1, y1, x3, s| (s * (((x1 - x3) + secp_modulus) % secp_modulus) - y1 + secp_modulus) % secp_modulus;
+
+    // Note that the most significant limb is 32-Bit instead of 16
+    let compute_q0_for_eq1 = |x1, y1, x2, y2, s, limb_index| ((-(s * x2 - s * x1 - y2 + y1) / secp_modulus + (1 << 258)) >> (limb_index * 16)) & if limb_index < 15 { 0xffff } else { 0xffffffff };
```

Comment:
Can you split those up a bit? I cannot really parse it.

Reply:
Yeah, good call! I made a `select_limb(x, limb_index)` function, and actually pulled it out all the way to the actual hint. This has the extra benefit that the evaluation cache will be *much* more efficient, because e.g. `compute_q0_for_eq1` won't have a `limb_index` argument anymore!